### PR TITLE
DHFPROD-1981: Smart Mastering improvements

### PR DESCRIPTION
**Pull request by ryanjdew**

### CODE REVIEW

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/extensions/runFlow.sjs [hunk 1]
```diff
@@ -16,30 +16,41 @@
...
-    let targetDatabase = params["target-database"] ? xdmp.database(params["target-database"]) : xdmp.database(datahub.config.FINALDATABASE);
-
+    let stepRef = flow.steps[stepNumber];
+    let stepDetails = datahub.flow.step.getStepByNameAndType(stepRef.stepDefinitionName, stepRef.stepDefinitionType);
+    let flowOptions = flow.options || {};
+    let stepRefOptions = stepRef.options || {};
+    let stepDetailsOptions = stepDetails.options || {};
+    // build combined options
+    let combinedOptions = Object.assign({}, flowOptions, stepDetailsOptions, stepRefOptions, options);
+    // combine all collections
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/extensions/runFlow.sjs]: I see differences with the 'options' precedence in different places.

collector.sjs -> let combinedOptions = Object.assign({}, baseStep.options, stepDoc.options, flowDoc.options, options);  - 'options' 's precedence is different. I am using this order for WriteStepRunner on the client side

impl/flow.sjs ->    let combinedOptions = Object.assign({}, stepDetails.options, stepRef.options, options); - flow options not used

Comment (ryanjdew) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/extensions/runFlow.sjs]: Good catch. We should definitely be consistent. The way I see it the precedence should be 1. options passed, 2. stepRef/baseStep (inside flow), 3. flow options, and 4. stepDetails/stepDoc.

@aebadirad, is that the way you see it?

Comment (aebadirad) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/extensions/runFlow.sjs]: yes sir.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/flow.sjs [hunk 2]
```diff
@@ -173,7 +173,7 @@ class Flow {
     let stepDetails = this.step.getStepByNameAndType(stepRef.stepDefinitionName, stepRef.stepDefinitionType);
 
     //here we consolidate options and override in order of priority: runtime flow options, step defined options, process defined options
-    let combinedOptions = Object.assign({}, stepDetails.options, stepRef.options, options);
+    let combinedOptions = Object.assign({}, stepDetails.options, flow.options, stepRef.options, options);
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/flow.sjs]: There is another place where we combine options line 262 in runStep() method in impl/flow.sjs

Comment (ryanjdew) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/flow.sjs]: That should be removed. At that point, the options have already been combined.

**Pull request merged**